frame/pablo/src/lib.rs

```diff
@@ -467,6 +499,139 @@ pub mod pallet {
 		pub(crate) fn account_id(pool_id: &T::PoolId) -> T::AccountId {
 			T::PalletId::get().into_sub_account(pool_id)
 		}
+		pub(crate) fn do_get_exchange_rate(
+			pool_id: T::PoolId,
+			swap: bool,
+		) -> Result<Rate, DispatchError> {
+			let pair = Self::currency_pair(pool_id)?;
+			let pool_account = Self::account_id(&pool_id);
+			let pair = if swap { pair.swap() } else { pair };
+			let pool_base_asset_under_management =
+				T::Convert::convert(T::Assets::balance(pair.base, &pool_account));
+			let pool_quote_asset_under_management =
+				T::Convert::convert(T::Assets::balance(pair.quote, &pool_account));
+			Ok(Rate::checked_from_rational(
+				pool_base_asset_under_management,
+				pool_quote_asset_under_management,
+			)
+			.ok_or(ArithmeticError::Overflow)?)
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_n: T::BlockNumber) -> Weight {
+			let current_timestamp = T::Time::now();
+			Pools::<T>::iter()
+				.filter(|(pool_id, pool)| {
+					let pair = match pool {
+						PoolConfiguration::StableSwap(stable_swap_pool_info) =>
+							stable_swap_pool_info.pair,
+						PoolConfiguration::ConstantProduct(constant_product_pool_info) =>
+							constant_product_pool_info.pair,
+						PoolConfiguration::LiquidityBootstrapping(
+							liquidity_bootstrapping_pool_info,
+						) => liquidity_bootstrapping_pool_info.pair,
+					};
+					T::EnabledPoolTWAP::contains(&(pair, *pool_id))
```

Comment:
This filter still requires iterating over all pools. Why not iterate over `T::EnabledPoolTWAP` to avoid needless storage access?

Reply:
Yeah I agree this was originally done with permissioned creation in mind, meaning the number of pool would be very limited. Let's make it such that `EnabledPoolTWAP` is the iterator

Reply:
Yeah I agree this was originally done with permissioned creation in mind, meaning the number of pool would be very limited. Let's make it such that `EnabledPoolTWAP` is the iterator

Reply:
Okay